cmd/api-router-cors.go

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2015-2024 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+
+	xhttp "github.com/minio/minio/internal/http"
+	"github.com/minio/mux"
+	xcors "github.com/minio/pkg/v3/cors"
+	"github.com/minio/pkg/v3/wildcard"
+	"golang.org/x/exp/slices"
+
+	rscors "github.com/rs/cors"
+)
+
+// There are two ways to apply CORS in MinIO:
+// 1. Global CORS settings: This was the original method that applied to the entire API surface,
+//    and was limited to just allowed origins.
+// 2. Bucket-specific CORS settings: This is a new method that allows for more fine-grained
+//    control over CORS settings, and can be applied on a per-bucket basis. It uses AWS's own spec
+//    that allows a number of rules to be specified. Compatible with the global settings, takes precedence
+//    https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html
+
+// corsHandler returns a middleware that is able to either apply bucket-specific CORS settings if they exist,
+// or fallback to the global CORS settings if they don't.
+func corsHandler(handler http.Handler) http.Handler {
+	corsGlobal := corsGlobalHandler(handler)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		bucket := vars["bucket"]
+		if bucket == "" {
+			// No bucket, continue the request using global CORS.
+			corsGlobal.ServeHTTP(w, r)
+			return
+		}
+
+		// Check if the bucket has a CORS config set.
+		config, _, err := globalBucketMetadataSys.GetCorsConfig(bucket)
```

Comment:
can you wrap this around singleflight package so that we can de-duplicate concurrent callers.

Reply:
@harshavardhana  I expect the majority of calls to GetCorsConfig to just be a map lookup, with read concurrency supported via RLock here: https://github.com/minio/minio/blob/master/cmd/bucket-metadata-sys.go#L449

Consider that singleflight will lock a mutex here: https://cs.opensource.google/go/x/sync/+/master:singleflight/singleflight.go;l=92,  so if we use it in the wrong place, we restrict the concurrency unnecessarily.

What are we protecting? Maybe we just want to protect the scenario where we have a cache miss (i.e. metadataMap[bucket] not found), here: https://github.com/minio/minio/blob/master/cmd/bucket-metadata-sys.go#L454

In that case I would add the singleflight at that location in `BucketMetadataSys.GetConfig()` and then it would apply to all requests for bucket config that are cache misses.

Let me know your thoughts

Reply:
We are not protecting we are simply deduplicating the call.